airflow/providers/pinecone/operators/pinecone_operator.py

```diff
@@ -0,0 +1,83 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from __future__ import annotations
+
+from functools import cached_property
+from typing import TYPE_CHECKING, Any, Sequence
+
+from airflow.models import BaseOperator
+from airflow.providers.pinecone.hooks.pinecone_hook import PineconeHook
+
+if TYPE_CHECKING:
+    from airflow.utils.context import Context
+
+
+class PineconeIngestOperator(BaseOperator):
+    """
+    Airflow Operator to ingest vector embeddings into Pinecone.
+
+    .. seealso::
+        For more information on how to use this operator, take a look at the guide:
+        :ref:`howto/operator:PineconeIngestOperator`
+
+    :param conn_id: pinecone_conn_id: The connection id to use when connecting to Pinecone.
+    :param index_name: Name of the Pinecone index.
+    :param vectors: Data to be ingested, in the form of a list of tuples where each tuple
+        contains (id, vector_embedding, metadata).
+    :param namespace: The namespace to write to. If not specified, the default namespace is used.
+    :param batch_size: The number of vectors to upsert in each batch.
+    """
+
+    template_fields: Sequence[str] = ("input_vectors",)
+
+    def __init__(
+        self,
+        *,
+        conn_id: str = PineconeHook.default_conn_name,
+        index_name: str,
+        input_vectors: list[Any],
+        namespace: str = "",
+        batch_size: int | None = None,
+        **kwargs: Any,
+    ) -> None:
+        self.upsert_kwargs = kwargs.pop("upsert_kwargs", {})
```

Comment:
Why not add `upsert_kwargs ` as a parameter for the operator?

Reply:
+1. It would be better to be explicit than implicit here so folks don't need to read source code to find these hidden kwargs.

Reply:
Same thought - https://github.com/apache/airflow/pull/35023#discussion_r1376207422

Reply:
Added upsert_kwargs as a param